File: backend/src/main/java/com/controleentregas/service/LocalidadeService.java
package com.controleentregas.service;

import com.controleentregas.domain.entity.Localidade;
import com.controleentregas.dto.request.LocalidadeRequest;
import com.controleentregas.dto.response.LocalidadeResponse;
import com.controleentregas.dto.response.PageResponse;
import com.controleentregas.exception.RecursoNaoEncontradoException;
import com.controleentregas.repository.LocalidadeRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class LocalidadeService {

    private final LocalidadeRepository repository;

    public LocalidadeService(LocalidadeRepository repository) {
        this.repository = repository;
    }

    @Transactional(readOnly = true)
    public PageResponse<LocalidadeResponse> listar(String busca, Pageable pageable) {
        Page<Localidade> page = (busca != null && !busca.isBlank())
                ? repository.findByEnderecoContainingIgnoreCase(busca, pageable)
                : repository.findAll(pageable);
        return PageResponse.de(page.map(LocalidadeResponse::de));
    }

    @Transactional(readOnly = true)
    public LocalidadeResponse buscarPorId(Long id) {
        return LocalidadeResponse.de(buscarEntidade(id));
    }

    @Transactional
    public LocalidadeResponse criar(LocalidadeRequest request) {
        Localidade localidade = new Localidade();
        aplicarDados(localidade, request);
        return LocalidadeResponse.de(repository.save(localidade));
    }

    @Transactional
    public LocalidadeResponse atualizar(Long id, LocalidadeRequest request) {
        Localidade localidade = buscarEntidade(id);
        aplicarDados(localidade, request);
        return LocalidadeResponse.de(repository.save(localidade));
    }

    @Transactional
    public void excluir(Long id) {
        Localidade localidade = buscarEntidade(id);
        repository.delete(localidade);
    }

    private Localidade buscarEntidade(Long id) {
        return repository.findById(id)
                .orElseThrow(() -> RecursoNaoEncontradoException.porId("Localidade", id));
    }

    private void aplicarDados(Localidade localidade, LocalidadeRequest request) {
        localidade.setEndereco(request.endereco());
        localidade.setTipoVia(request.tipoVia());
        localidade.setAptidaoChuva(request.aptidaoChuva());
        localidade.setObservacoes(request.observacoes());
    }
}

File: frontend/src/services/localidadeService.ts
import api from './api';
import { PageResponse } from '../types/api';
import { Localidade, LocalidadeRequest } from '../types/entities';

export const localidadeService = {
  listar: (params: { page?: number; size?: number; busca?: string } = {}) =>
    api.get<PageResponse<Localidade>>('/localidades', {
      params: { page: params.page ?? 0, size: params.size ?? 10, ...(params.busca ? { busca: params.busca } : {}) },
    }).then((r) => r.data),

  buscarPorId: (id: number) =>
    api.get<Localidade>(`/localidades/${id}`).then((r) => r.data),

  criar: (data: LocalidadeRequest) =>
    api.post<Localidade>('/localidades', data).then((r) => r.data),

  atualizar: (id: number, data: LocalidadeRequest) =>
    api.put<Localidade>(`/localidades/${id}`, data).then((r) => r.data),

  excluir: (id: number) =>
    api.delete<void>(`/localidades/${id}`).then((r) => r.data),
};
